### ISSUE TITLE
prop types rule not working with hoc

### ISSUE THREAD

**Issue by krzysztofpniak** [yannickcr/eslint-plugin-react#2931]
Following code validation runs without any errors:
```jsx
const withDebug = name => Component => {
  const Wrapped = forwardRef(({scope, ...rest}, ref) => {
    useEffect(() => {
      console.log(name, 'mount');
      return () => {
        console.log(name, 'unmount');
      };
    }, []);

    return <Component ref={ref} {...rest} />;
  });

  Wrapped.propTypes = {
    scope: PropTypes.string.isRequired,
  };

  return Wrapped;
};

const Test = withDebug('test')(({text}) => <div>{text}</div>);
```

I'm pretty sure that eslint-plugin-react should return error that _Test_ component requires definition of _text_ property type.

**Comment by ljharb** [comment]
If `Wrapped` is a component, its second argument wouldn't be `ref`, it would be `context`. The `ref` argument is only for `forwardRef` callbacks, which doesn't apply to this snippet.

**Comment by krzysztofpniak** [comment]
@ljharb you're right. I fixed the code above.

**Comment by ljharb** [comment]
I’m not sure if this plugin reports on forwardRef components at all, tbh, since they’re a bit of a special case.

**Comment by ljharb** [comment]
This is pretty metaprogrammy, so there's not really a way this can work with static analysis. I'd make the component on the last line not be inline, and then we'll be able to detect it statically.
